Add page render functionality

tameTNT opened the pull request:
The page thumbnail generation function is finally complete (I think/hope!).
A lot of things have been tweaked across models and templates, although I was merging with some of the other active branches during the process to try and reduce the change of conflicts now when the other branches are also merged into master eventually.
Please, please **get a local copy of this branch and dig out anything that might not be working**.
Since this is such a big set of changes, we should probably have achieved universal agreement on everything in Discord before merging this branch into master.

tameTNT committed:
Install imgkit

imgkit module installed via pipenv.
Pipfile and Pipfile.lock updated accordingly.

tameTNT committed:
Change use of os to Path module

Path module use is more pythonic and recommended by code jam organisers
over using os.join().

tameTNT committed:
Start create render function and set up template

README.md updated with new environment variable setup

tameTNT committed:
Merge remote-tracking branch 'origin/console-theme' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/artemis-dev' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-page_maker-app' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-base-views' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/django_meta/urls.py
#	jolly-jellyfish/src/page_maker/models.py

tameTNT committed:
Simplify and expand README.md

Refine wording of Introduction and Idea sections.
Add more line breaks for readability when editing README.md file.
Add explanation of environment variable config to setup section.
Add complete language name to code section syntax highlighting options.

tameTNT committed:
Update models.py and adapt templates accordingly.

Change models.py (and be extension forms.py) with some path changes and
use of RelativeFilePathField (pipenv install required) instead of
models.FilePathField for easier stylesheet access within templates.

Reload the database files too -
delete original db and migrations and start afresh.
Delete residual file: generic_template.html no longer required.

Edit base.html to be able to access stylesheet by static path.
Edit webpage_list.html and webpage_view.html with minor formatting
changes for when there are no sites and to actually load user images +
prevent errors when no images present.

tameTNT committed:
Add sys-env  warning to render_img_thumbnail.py

Add try-except block to cover if user has not added WKHTML_TO_IMAGE to
system environment variables as specified in the readme.

tameTNT committed:
Merge remote-tracking branch 'origin/master' into add-page-render-functionality

tameTNT committed:
Implement page thumbnail generation.

Thumbnail rendering function created as a custom templatetag.
This allows it to be executed from within a template when it is detected
that the page currently uses placeholder_img.png in its thumbnail field.

webpage_view.html also rearranged to make more logical sense.

tameTNT committed:
Simplify render_img_thumbnail.py tag functions

webpage_list.html adapted accordingly. Comments added to page code too.
More specificity added to models.py as well
to provide greater input freedom.
(typo corrected in views.py)

tameTNT committed:
Simplify render_img_thumbnail.py tag functions

webpage_list.html adapted accordingly. Comments added to page code too.
More specificity added to models.py as well
to provide greater input freedom.
(typo corrected in views.py)

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html

tameTNT committed:
Merge remote-tracking branch 'origin/master' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-base-views' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html
#	jolly-jellyfish/src/page_maker/views.py

tameTNT committed:
Improve and rework thumbnail image rendering

Thumbnails are now generated when the user reaches the page itself
(not the list page). This is accomplished within base.html.
This means thumbnails are generated immediately following page creation
(since user is directed to the actual webpage after filling in form).

Thumbnails are also automatically regenerated when a page is edited.

*However, as noted in the Discord chat, some underlying issues have been
discovered including background images (in css) not being loaded by
imgkit.
This is due to the underlying wkhtmltoimage binary used and hence
doesn't appear solvable using imgkit alone.
Any potential solutions - feel free to add.

Artemis21 committed:
Add global styling.

And extended version of the Windows 95 theme. Also added headers to pages as it selects the first element and styles it as a header.

Artemis21 committed:
Merge branch 'add-page-render-functionality' into add-global-style

Artemis21 committed:
Use Pathlib for paths for cross platform compability.

tameTNT committed:
Improve layout of 'best' and 'user' pages

Thumbnails and page links are now tabulated.
This makes the page more compact as well as easier to use.
(In prep. for standardised thumbnail image size
this make pages more uniform)

tameTNT committed:
Upgrade render_img_thumbnail.py to use Selenium

Selenium offers a major functional improvement over the previous imgkit.
Although it may be slightly slower (probs. due to browser load).

Update Pipfile and .lock with the new package requirements.
Update README.md with new download and env variable requirements.
Update models.py path gen function for Template.style_sheet due to newly
added site-wide style.css which should not be accessible
for user template creation.
Change affected stylesheet paths used in base.html

tameTNT committed:
Upgrade render_img_thumbnail.py to use Selenium

Selenium offers a major functional improvement over the previous imgkit.
Although it may be slightly slower (probs. due to browser load).

Update Pipfile and .lock with the new package requirements.
Update README.md with new download and env variable requirements.
Update models.py path gen function for Template.style_sheet due to newly
added site-wide style.css which should not be accessible
for user template creation.
Change affected stylesheet paths used in base.html

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

tameTNT committed:
Add Template preview support to thumbnail rendering

Add extended documentation and annotation to render_thumbnail.py.
Add new TemplateView to preview what a stylesheet looks like.

Selenium can now be run using either Firefox or Chromium.

tameTNT committed:
Update README.md with Firefox driver detail

Added readme information concerning use of the geckodriver for Firefox.

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+
 ## Initial Setup
-```sh
+### wkhtmltoimage
````
Did you forget to update this heading?

tameTNT replied:
It appears so!

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+(venv) $ python manage.py runserver
+```
+
+[1]: https://sites.google.com/a/chromium.org/chromedriver/home
+[2]: https://github.com/mozilla/geckodriver/releases
````
Could we move these footnotes up? (Don't really know much about markdown footnotes :p)... It's annoying you have to scroll so far down to get context in plaintext.

tameTNT replied:
In the actual README these links are obviously clickable in their original location so it's not too much of an issue where these footnotes are.

A5rocks commented on jolly-jellyfish/README.md:
```diff
@@ -1,18 +1,52 @@
-$ pipenv run makemigrations  # TODO: this might be optional.
-$ pipenv run migrate
+$ pipenv shell
+(venv) $ python manage.py makemigrations  # TODO: this might be optional, depends on pipenv script section.
```
My intent by "this might be optional" is that the migrations might already be made for the final application, meaning we don't have to run `makemigrations`.

tameTNT replied:
Oh ok - I get you.
```suggestion
(venv) $ python manage.py makemigrations  # TODO: this might be optional, depends on migrations already run.
```

A5rocks commented on jolly-jellyfish/README.md:
```diff
@@ -1,18 +1,52 @@
-$ pipenv run migrate
+$ pipenv shell
+(venv) $ python manage.py makemigrations  # TODO: this might be optional, depends on pipenv script section.
+(venv) $ python manage.py migrate
```
Don't forget to `cd src` first! (or specify your directory before the shell session)

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
-$ pipenv run start
-```
\ No newline at end of file
+```shell script
````
Should we use `sh` for syntax highlighting on GitHub?

tameTNT replied:
```suggestion
```sh
(venv) $ cd src
```

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
+
+### pipenv
+```shell script
````
Should we use `sh` for syntax highlighting on GitHub?

tameTNT replied:
Yeah, I think this was just pycharm autocompleting it for me.

A5rocks commented on jolly-jellyfish/README.md:
```diff
@@ -1,18 +1,52 @@
+In a production environment, this process would obviously not be needed as the server would be running a known OS.
+Therefore, only one executable would need to be download to the src directory and then hard linked within the code; rather than having to use `os.environ`.
+
+Please also note that, as described within `src/page_maker/templatetags/render_thumbnail.py`, 
+asynchronous tasks (using `Celery` and `Redis`) are not easily setup on Windows - 
+hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
```
Could we drop a `FIXME: find if we can use docker for celery / redis` before this (I think PyCharm will catch that) so that we can see if there's a way before finishing the project?

tameTNT replied:
Baru did mention potentially setting one up in the Discord.

A5rocks commented on jolly-jellyfish/README.md:
```diff
@@ -1,18 +1,52 @@
+asynchronous tasks (using `Celery` and `Redis`) are not easily setup on Windows - 
+hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
+
+### pipenv
```
```suggestion
### Using pipenv
```

Then we can add a `using <xyz>` if we want to use something else.

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
-```
\ No newline at end of file
+```shell script
+(venv) $ python manage.py runserver
````
Don't forget to `cd src` or specify the directory you are in.

A5rocks commented on jolly-jellyfish/src/.flake8:
```diff
@@ -0,0 +1,4 @@
+[flake8]
+exclude = */migrations/*
```
I'm pretty sure the organizers did this for us in the root level flake8 config. Maybe we can remove this line once we confirm it works without?

https://github.com/python-discord/summer-code-jam-2020/commit/4419d74ebc88a7538721268af1d6f464735c3ebe

A5rocks replied:
I'll be fine with doing this on master eventually.

A5rocks commented on jolly-jellyfish/src/django_meta/settings.py:
```diff
@@ -4,9 +4,14 @@
+MEDIA_URL = "/user_media/"
+
+LOGIN_REDIRECT_URL = '/'
+LOGIN_URL = '/users/login'
```
Could we be consistent on `"` vs `'`?

A5rocks commented on jolly-jellyfish/src/django_meta/urls.py:
```diff
@@ -1,7 +1,13 @@
+    path('users/login/', auth_views.LoginView.as_view(), name='login'),
+    path('users/logout/', auth_views.LogoutView.as_view(next_page='/'), name='logout'),
+    path('', include('page_maker.urls')),
+] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
Did `flake8` really let this file have no line ending?

tameTNT replied:
```suggestion
] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)

```

A5rocks commented on jolly-jellyfish/src/.flake8:
```diff
@@ -0,0 +1,4 @@
+[flake8]
+exclude = */migrations/*
+ignore = D,I,ANN,F
```
TBH this feels a little too generic... Could you put comments specifying why we're ignoring each one?

tameTNT replied:
I don't actually know what this is - this flake8 file was from @Artemis21 's branch I think.

Artemis21 replied:
Oh, I ignored those because the judges are running with no extensions and those were all extensions I had installed. I should probably just have ignored them in the command.

A5rocks commented on jolly-jellyfish/src/django_meta/urls.py:
```diff
@@ -1,7 +1,13 @@
     path('admin/', admin.site.urls),
-]
+    path('users/login/', auth_views.LoginView.as_view(), name='login'),
+    path('users/logout/', auth_views.LogoutView.as_view(next_page='/'), name='logout'),
```
Just wondering (this though popped up into my head), do we have any specific strategy to determine whether a url should end with `/`?

tameTNT replied:
Django recommends trailing slashes (and by default they are auto-appended to the end of URLs without them) - https://stackoverflow.com/questions/42212122/why-django-urls-end-with-a-slash

A5rocks commented on jolly-jellyfish/src/page_maker/forms.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class WebpageForm(forms.ModelForm):
+
```
Unnecessary spacing, and it's debatable if this helps increase code legibility.

tameTNT replied:
See above

A5rocks commented on jolly-jellyfish/src/page_maker/forms.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class CommentForm(forms.ModelForm):
+
```
Unnecessary spacing, and it's debatable if this helps increase code legibility.

tameTNT replied:
This spacing is convention since 'local' imports (.models) are separated from package imports and there should always be two lines between class (and function) definitions.

Artemis21 replied:
Yeah IMO this should stay as is.

A5rocks commented on jolly-jellyfish/src/page_maker/forms.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class TemplateForm(forms.ModelForm):
+
```
Unnecessary spacing, and it's debatable if this helps increase code legibility.

tameTNT committed:
Clarify template logic and fix url inconsistency

The new user_detail.html and webpage_list.html account for the default
value of a Webpage object - with 'placeholder_img' in the path -
not just a blanket check for the url (which will always be True).

For Template-object-related url templates,
one url used the pk while another used the templatename.
Both now use templatename.
(since this field is now unique in the Template model)

A5rocks commented on jolly-jellyfish/src/page_maker/migrations/0001_initial.py:
```diff
@@ -1,12 +1,15 @@
-# Generated by Django 3.0.8 on 2020-08-01 12:33
+# Generated by Django 3.0.9 on 2020-08-03 15:27
```
Us changing Django versions might be troublesome...

Artemis21 replied:
[Django aims to be as backward compatible as possible](https://docs.djangoproject.com/en/3.1/misc/api-stability/), I don't think it will be too much of an issue

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
 
 def get_theme_path():
-    return os.path.join(settings.BASE_DIR, 'static')
+    return str(Path(settings.BASE_DIR) / 'static' / 'themes')
```
We might not need this `str`... Could we try without it / `(Path(settings.BASE_DIR) / 'static' / 'themes')).resolve()` if the first doesn't work...

This might make the behavior easier to reason out.

tameTNT replied:
The RelativeFilePathField that uses this function below, I'm pretty sure only takes strings for the file path - it raised an error when it was provided a Path() object during testing.

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
+    #  potentially upload to user_themes within static not media? -
+    #  makes finding easier below in RelativeFilePathField
     # style_sheet = models.FileField(null=False, blank=False,
     #                      upload_to='themes/', validators=[FileExtensionValidator(['css'])])
```
I don't think is necessary given that we have a `style_sheet` field that's uncommented below. Though if this is a potential thing for a feature then maybe add a comment above about why it's there.

tameTNT replied:
```suggestion
```
Agreed - the todo comment sums up what's necessary - we don't need the commented out attribute.

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
+    # model.FilePathField() does not have a url attribute,
+    # so accessing the files within a html template is awkward.
+    # Using RelativeFilePathField allows for using django static support easily within the template
     # match uses regex applied to base filename only(.*=any characters, \.=., $=end of string)
```
We might want `.+` instead of `.*` to prevent someone from uploading a file called `.css`.

tameTNT replied:
Yep - I'm not very experienced with Regex but that sounds good!

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
+
+    # null=False and blank=False enforce there always being a thumbnail image associated
+    thumbnail = models.ImageField(null=False, blank=False,
+                                  verbose_name="thumbnail (leave empty-generated automatically)",
```
Instead of saying "leave empty", is there a way we can have it not appear in the admin site?

tameTNT replied:
```suggestion
    thumbnail = models.ImageField(null=False, blank=False, editable=False
                                  verbose_name='thumbnail (leave empty-generated automatically)',
```

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
+    # null=False and blank=False enforce there always being a thumbnail image associated
+    thumbnail = models.ImageField(null=False, blank=False,
+                                  verbose_name="thumbnail (leave empty-generated automatically)",
+                                  default='thumbnails/placeholder_img.png', upload_to='thumbnails/',
```
Choose between `'` and `"` :P

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
-    thumbnail = models.ImageField(null=True, blank=True, default=None,
-                                  upload_to='thumbnails/',
+    thumbnail = models.ImageField(null=False, blank=False,
+                                  verbose_name="thumbnail (leave empty-generated automatically)",
```
Could we have it not appear in the admin site instead of saying `leave empty-generated automatically`?

tameTNT replied:
```suggestion
    thumbnail = models.ImageField(null=False, blank=False, editable=False,
                                  verbose_name='thumbnail (leave empty-generated automatically)',
```

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
 
 
 class Webpage(models.Model):
```
I noticed `Webpage` and `Template` share some fields like `name` and `thumbnail`. Is it possible to make a pseudo-Model that has these fields and then subclass for `Webpage` and `Template`?

tameTNT replied:
I'm not sure - @baru64 and @Artemis21 might know more.

Artemis21 replied:
To be honest I still don't see the point of having Template separate from Webpage, but yes, IIRC Django models support inheritance. It might end up making an empty table for the base class though...

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
-                                  upload_to='thumbnails/',
+    thumbnail = models.ImageField(null=False, blank=False,
+                                  verbose_name="thumbnail (leave empty-generated automatically)",
+                                  default='thumbnails/placeholder_img.png', upload_to='thumbnails/',
```
Choose between `'` and `"` :P

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
+                                  default='thumbnails/placeholder_img.png', upload_to='thumbnails/',
                                   validators=[validate_image_file_extension])
+    # Implausible date is just to signify it does not yet have a thumbnail
+    thumbnail_edit_date = models.DateTimeField(default=datetime(1, 1, 1), editable=False)
```
Is there a way to just make this field Optional (or nullable or something) so we don't have to resort to hacks like this?

tameTNT replied:
I need it to be comparable to a DateTime object (e.g. in thumbnail_edit_date < edit_date) so None isn't sufficient as None can't be compared.

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
 
+    def save(self, *args, **kwargs):
+        """Trigger timestamp updates when object is edited"""
+        if not self.id:
```
Could we check `self.date_created` instead of `self.id`? Just so it looks less hacky...

tameTNT replied:
This is just the standard form I belive for checking if this is an object's initial creation save call. Could be commented for clarity I guess.
```suggestion
        if not self.id:  # if initial object instantiation
```

A5rocks commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -31,43 +44,48 @@ def get_user_dir_path(instance, filename):
+        if not self.id:
+            self.date_created = timezone.now()
+        self.last_edit_date = timezone.now()
+        return super(Webpage, self).save(*args, **kwargs)
```
```suggestion
        return super().save(*args, **kwargs)
```

Python 3 thing!

tameTNT replied:
Is it that simple? 😆 Whoops!

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/base.html:
```diff
@@ -0,0 +1,58 @@
+    <a href="{% url 'login' %}">Login</a>
+    <a href="{% url 'register' %}">Register</a>
+    {% endif %}
+    <a href="{% url 'home' %}">Home</a>
```
Should we really be putting this into the base template?

tameTNT replied:
Yep, the 'user login' interface should be consistent across pages on the site - i.e. accessible everywhere.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+  <form method="POST" enctype="multipart/form-data">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Create template</button>
+    </div>
```
Why do we have a wrapping div? (Just wondering...)

tameTNT replied:
Might be a stylesheet thing - @Artemis21 ?

Artemis21 replied:
No, as long as there is a header element at the top the stylesheet is happy.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html:
```diff
@@ -0,0 +1,15 @@
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <p>Delete template <b>{{template.name}}</b></p>
+      <button type="submit">Confirm</button>
+    </div>
```
Why do we have a wrapping div? (Just wondering...)

tameTNT replied:
Same as above @Artemis21

Artemis21 replied:
As above - I didn't add these.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html:
```diff
@@ -0,0 +1,15 @@
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <p>Delete account <b>{{user.username}}</b></p>
+      <button type="submit">Confirm</button>
+    </div>
```
Why do we have a wrapping div?

Artemis21 replied:
As above

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html:
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
```
I noticed this page shares similarity with `template_create`, `template_delete`, etc. Should we make a `form` template so we can just do:

```html
{% entends "page_maker/form.html" %}

{% block title %}
Delete page
{% endblock %}

{% block info %}
<p>Delete account <b>{{user.username}}</b></p>
{% endblock %}
```

Artemis21 replied:
That sounds like a good idea.

Artemis21 replied:
Anything happen to this?

A5rocks replied:
It's out of the scope for the PR, I noted down the idea for another PR

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -0,0 +1,27 @@
+{% endif %}
+<div>
+  <table>
+    {% for webpage in user_pages %}
```
Why are we explicitly passing in `user_pages`? Is there a `viewed_user.pages` thing?

tameTNT replied:
This page only displays the user's own pages (i.e. the ones they've created - it's like their showreel). In the class view, 'user_pages' forms part of the context given to this template.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -0,0 +1,27 @@
+<a href="{% url 'user-delete' pk=viewed_user.id %}">delete</a>
+{% endif %}
+<div>
+  <table>
```
This table might benefit from a `<th>` element! (just check out the first example on https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table)

tameTNT replied:
Agreed!

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -0,0 +1,27 @@
+        <tr>
+        <td>
+        {% if "placeholder_img" not in webpage.thumbnail.url %}
+            <img src="{{ webpage.thumbnail.url }}">
```
Can we wrap the `img` in an `a` so you can click the image to go to the page?

tameTNT replied:
I like that - although, in fairness, the link to the page is right next to it.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -0,0 +1,27 @@
+<a href="{% url 'user-update' pk=viewed_user.id %}">update</a>
+<a href="{% url 'user-delete' pk=viewed_user.id %}">delete</a>
+{% endif %}
+<div>
+  <table>
+    {% for webpage in user_pages %}
+        <tr>
+        <td>
+        {% if "placeholder_img" not in webpage.thumbnail.url %}
+            <img src="{{ webpage.thumbnail.url }}">
+        {% else %}
+            Go to page itself to generate a thumbnail for it.
+        {% endif %}
+        </td>
+        <td>
+            <a href="{% url 'webpage-view' pagename=webpage.name %}">{{ webpage.user_title }}</a>
+        </td>
+        </tr>
+    {% endfor %}
+    </table>
+</div>
```
Why do we have this random `div` by the way? Styling?

tameTNT replied:
Again 😆 , @Artemis21 ?

Artemis21 replied:
Again, no idea, didn't add it

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Update</button>
+    </div>
```
Another random `div`... Is this for styling?

Artemis21 replied:
As above, no

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html:
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
```
Check out my comment on `page_maker/user_delete.html`

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_create.html:
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h2>Create page</h2>
```
```suggestion
<h2>Create your page!</h2>
```

Put some energy into it :)

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html:
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
```
See my comment on `user_delete`

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html:
```diff
@@ -0,0 +1,22 @@
+<div><h1>{{ webpage.user_title }}</h1></div>
+<div>
+  {% for comment in comments %}
+  <div>
+    <p>{{ comment.author.username }}: {{ comment.title }}</p>
+    <p>{{ comment.content }}</p>
+  </div>
```
Can we separate the comments by putting them in a `table` or something?

tameTNT replied:
Good styling suggestion. @Artemis21

Artemis21 replied:
```suggestion
  <div class="comments">
    <p class="comment-header">{{ comment.author.username }}: {{ comment.title }}</p>
    <p class="comment-body">{{ comment.content }}</p>
  </div>
```

`table`s shouldn't be used for non-table layout, but I will certainly add some styling to comments - it would help if these `p`s and `div`s had classes.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html:
```diff
@@ -0,0 +1,22 @@
+        {{ comment_form.as_p }}
+        <button type="submit">Add comment</button>
+      </div>
+    </form>
```
1 too many indents...

tameTNT replied:
```suggestion
  <form method="POST" enctype="multipart/form-data" action="{% url 'comment-create' pagename=webpage.name %}">
    {% csrf_token %}
    <div>
      {{ comment_form.as_p }}
      <button type="submit">Add comment</button>
    </div>
  </form>
```
Indentation adjusted

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html:
```diff
@@ -0,0 +1,27 @@
+    {% for webpage in webpages %}
+        <tr>
+        <td>
+        {% if "placeholder_img" not in webpage.thumbnail.url %}
```
Should we make the thumbnail url optional so we don't have to rely on `"placeholder_img"` being in it?

tameTNT replied:
I think I considered this originally but, to be honest, I think this actually helps readability slightly by saying that we're only going to show a thumbnail if it's actually been generated already and isn't just the initial placeholder.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html:
```diff
@@ -0,0 +1,14 @@
+{% extends "page_maker/base.html" %}
```
See my comment on `user_delete`

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -0,0 +1,29 @@
+    <img src="{{ webpage.user_image_3.url }}" alt="User-uploaded footer image">
+    {% endif %}
+<p>
+    {{ webpage.user_text_3 }}
```
You aren't ending these tags :^)

tameTNT replied:
I think in like my first ever super-basic course of HTML like 6 years ago I just learnt not to close <p> tags because they didn't need it. My head seems to think of it more like a <br> rather than what it actually is which is a paragraph marker 😆 . The habit seems to have stuck around...

A5rocks commented on jolly-jellyfish/src/page_maker/templates/registration/login.html:
```diff
@@ -0,0 +1,25 @@
+ 
+         <input type="submit" value="login" />
+         <input type="hidden" name="next" value="{{ next }}" />
+     </form>
```
1 too many indents.

tameTNT replied:
```suggestion
        <table>
        <tr>
            <td>{{ form.username.label_tag }}</td>
            <td>{{ form.username }}</td>
        </tr>
        <tr>
            <td>{{ form.password.label_tag }}</td>
            <td>{{ form.password }}</td>
        </tr>
        </table>

        <input type="submit" value="login" />
        <input type="hidden" name="next" value="{{ next }}" />
      </form>
```

A5rocks commented on jolly-jellyfish/src/page_maker/templates/registration/register.html:
```diff
@@ -0,0 +1,18 @@
+{% extends "page_maker/base.html" %}
+{% block content %}
+   <div>
+       <h2>Sing Up</h2>
```
typo'd

tameTNT replied:
```suggestion
       <h2>Sign Up</h2>
```
It did sound fun though...

Artemis21 replied:
```suggestion
   <h2>Sign Up</h2>
   <div>
```
Typo as above, and styling requires that the header be the first element.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+from django.http import HttpRequest
+from django.utils import timezone
+from selenium import webdriver
+from selenium.webdriver.chrome.options import Options
```
Isn't this also available as `webdriver.ChromeOptions`?

tameTNT replied:
Thanks for picking this up actually - this messes with some of the stuff below. I'll change this in its own commit now.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+@register.simple_tag()
+def render_thumbnail(request: HttpRequest, url_to_render: str, page_obj: Union[Webpage, Template]) -> None:
+    """
+    This function needs to be called from within django html templates using:
+        {% load render_thumbnail %}
```
Could you add a space?

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+
+        {% render_thumbnail request url_to_render page_obj as none %}
+
+    It saves a thumbnail/screenshot of the page from within which it is called to page_obj's thumbnail attribute.
+    This is accomplished using a selenium browser driver running in a headless configuration.
+
+    Since, when running a local server, this results in the server being pinged twice (once within itself),
+    a ConnectionResetError may be output to the console.
+    This is safe to ignore (but can't be escaped, with except, for some reason).
```
FWIW, I would personally cap lines at 79 chars instead of at sentence ends. To each their own though.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+    a ConnectionResetError may be output to the console.
+    This is safe to ignore (but can't be escaped, with except, for some reason).
+
+    :param request: A Django HttpRequest object, sourced from the page which is calling this function.
+        This is needed in order to access additional GET info and to build a full url (including domain).
```
Isn't there a way to specify types instead of just saying `A Django HttpRequest object`?

tameTNT replied:
The HttpRequest object is specified in the function signature above.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+    :param page_obj: Either a 'Webpage' or 'Template' Django object. Accessed in order to save thumbnail path.
+    """
+
+    template_obj_bool = hasattr(page_obj, 'style_sheet')  # True if object is an instance of models.Template
```
Should we use `isinstance(page_obj, Template)`. Also, maybe rename this variable to `is_template` or something?

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+    outdated_bool = False
+    if not template_obj_bool:  # only Webpages have the thumbnail_edit_date attr (since they can be edited by the user)
+        # True if the thumbnail is outdated or does not exist yet (thumbnail_edit_date datetime set to 1, 1, 1)
+        outdated_bool = page_obj.thumbnail_edit_date < page_obj.last_edit_date
```
Noticed this now: maybe it should be called `thumbnail_create_date`, and we make new thumbnails instead of reusing the same objects? I don't know if that's worth it or not though.

tameTNT replied:
I like keeping it to the same objects - it also prevents thumbnail clutter within the directory where they're saved.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+    template_obj_bool = hasattr(page_obj, 'style_sheet')  # True if object is an instance of models.Template
+
+    outdated_bool = False
+    if not template_obj_bool:  # only Webpages have the thumbnail_edit_date attr (since they can be edited by the user)
```
Move this comment to the next line?

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+
+    # To avoid infinite loops as the page calls itself to render,
+    # ensures function only runs if url doesn't end with '?rendering=true' (added below)
+    rendering_bool = bool(request.GET.get('rendering', ''))
```
Maybe rename to `should_render` or `should_not_render` to make sure we know what it is going to be used for?

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+    rendering_bool = bool(request.GET.get('rendering', ''))
+
+    if (template_obj_bool or outdated_bool) and not rendering_bool:  # if a thumbnail is needed
+        base_dir = Path(settings.BASE_DIR)
+        media_dir = Path(settings.MEDIA_ROOT)
```
Should we store the settings as `Path`s to begin with?

tameTNT replied:
They're stored in the Django settings files and are created using os.join() by default. I don't want to risk changing something in the file which literally controls the entire project! Django uses these internally as well to piece together URLs so I think it's best left alone 😆

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+            window_size = '500,500'  # thumbnails for Templates, due to less content, can be much smaller
+        else:
+            window_size = '1980,1080'
+        OPTIONS.add_argument(f'--window-size={window_size}')
```
Should we make a new `options` objects so that we're not modifying something in global scope? (This may or may not be thread-safe, idk)

tameTNT replied:
This'll have to be changed anyway now that I realise Chrome and Firefox options aren't the same.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+        if 'chromedriver' in DRIVER_PATH:
+            driver = webdriver.Chrome(executable_path=DRIVER_PATH, options=OPTIONS)
+        elif 'geckodriver' in DRIVER_PATH:
+            driver = webdriver.Firefox(executable_path=DRIVER_PATH, options=OPTIONS)
```
I'm not _quite_ sure we can pass ChromeOptions into Firefox :^)

tameTNT replied:
Yep - see above!

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+        url = request.build_absolute_uri(url_to_render) + '?rendering=true'
+
+        # NB: This function raises: "ConnectionResetError: [WinError 10054] ...
+        # An existing connection was forcibly closed by the remote host" in console."
```
Weird ending quote.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+        # Celery (for asynchronous tasks) no longer supports Windows - same with Redis.
+        # This means that, in this current dev environment where there are multiple OSs at play,
+        # having a functioning webpage, while this function executes in the background,
+        # doesn't seem possible 😭
```
We might be able to use Docker?

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+        driver.close()
+
+        img = Image.open(full_os_path)
+        img.thumbnail(size=(500, 500))  # image shrunk to reduce storage space and reduce space used on actual website
```
FWIW I don't think a `1980x1080` image can be shrunk to `500x500` well :^)

tameTNT replied:
pillow takes those as max measurements - i.e. the image will be shrunk until the largest side is 500. Aspect ratio is maintained.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,116 @@
+        img.save(full_os_path)
+
+        page_obj.thumbnail = f'{image_path}'  # saves relative path to thumbnail to ImageField attribute
+        if not template_obj_bool:  # only Webpage objects have this thumbnail_edit_date attribute
```
You might want to instead comment that `only Webpage objects might edited after this and so have thumbnail_edit_date`

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+    path('users/<int:pk>/delete', UserDeleteView.as_view(), name='user-delete'),
+
+    path('pages/new', WebpageCreateView.as_view(), name='webpage-create'),
+    path('pages/<str:pagename>/', WebpageView.as_view(), name='webpage-view'),
```
Why are we ending this with `/`?

tameTNT replied:
```suggestion
    path('users/<int:pk>/update/', UserUpdateView.as_view(), name='user-update'),
    path('users/<int:pk>/delete/', UserDeleteView.as_view(), name='user-delete'),

    path('pages/new/', WebpageCreateView.as_view(), name='webpage-create'),
    path('pages/<str:pagename>/', WebpageView.as_view(), name='webpage-view'),
```

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+urlpatterns = [
+    path('', MainView.as_view(), name='home'),
+    path('users/register', UserRegister.as_view(), name='register'),
+    path('users/<int:pk>/', UserDetailView.as_view(), name='user-detail'),
```
Why are we ending this with `/`?

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+    path('comments/<int:pk>/delete', CommentDeleteView.as_view(), name='comment-delete'),
+
+    path('templates/new', TemplateCreateView.as_view(), name='template-create'),
+    path('templates/<str:templatename>/', TemplateView.as_view(), name='template-view'),
```
Why are we ending this with `/`?

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+    path('pages/<str:pagename>/detail', WebpageDetailView.as_view(), name='webpage-detail'),
+    path('pages/<str:pagename>/update', WebpageUpdateView.as_view(), name='webpage-update'),
+    path('pages/<str:pagename>/delete', WebpageDeleteView.as_view(), name='webpage-delete'),
+    path('pages/', WebpageListView.as_view(), name='webpage-list'),
```
Why are we ending this with `/`?

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        # user = get_object_or_404(User, pk=self.kwargs.get('pk'))
```
Useless commented line?

tameTNT replied:
I'm not sure, @baru64 might be reminding himself of something. It's also possible that this is already removed on his branch.

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    fields = ['first_name', 'last_name', 'email']
+    success_url = '/'
+
+    def test_func(self):
```
We might want to rename this function if we can... If not, move the functionality to anther function, call it something like `request_user_is_self` (bonus points if it is a property), then call it from this?

tameTNT replied:
This is actually an overriding method thing with Django and so cannot be renamed.
![image](https://user-images.githubusercontent.com/30503695/89539988-5f5ace00-d7f4-11ea-812d-1dd15267f380.png)

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    template_name = 'page_maker/user_delete.html'
+    success_url = '/'
+
+    def test_func(self):
```
See my comment above.

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+
+class WebpageView(DetailView):
+    """
+    Display webpage
+    """
```
Not exactly a needed docstring :^)

tameTNT replied:
Better it exists than it doesn't
```suggestion
    """
    Display user-created webpage itself (with links to User's page and comments, i.e. 'WebpageDetailView')
    """
```
Better? 😆

Artemis21 replied:
Yeah, I often put docstrings in *every* class and function even if it is obvious, some linters even require it

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    model = Webpage
+    template_name = 'page_maker/webpage_update.html'
+
+    def test_func(self):
```
See my comments on the previous `test_func`s

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    model = Webpage
+    template_name = 'page_maker/webpage_delete.html'
+
+    def test_func(self):
```
See my comments on the previous `test_func`s

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+class WebpageDeleteView(LoginRequiredMixin, UserPassesTestMixin, DeleteView):
+    model = Webpage
+    template_name = 'page_maker/webpage_delete.html'
```
PATHLIB

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+class WebpageUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
+    model = Webpage
+    template_name = 'page_maker/webpage_update.html'
```
PATHLIB

tameTNT replied:
I think this has to be a string actually and is how Django sets it out themselves too (https://docs.djangoproject.com/en/3.0/topics/class-based-views/generic-display/#generic-views-of-objects). It's not actually used as a path I think but as sort of a key to finding the correct template using Django's inner magic 👍 .

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+
+class TemplateCreateView(LoginRequiredMixin, FormView):
+    template_name = 'page_maker/template_create.html'
```
PATHLIB

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    Display template using example page
+    """
+    model = Template
+    template_name = 'page_maker/example_page.html'
```
PATHLIB

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+class TemplateDeleteView(LoginRequiredMixin, UserPassesTestMixin, DeleteView):
+    model = Template
+    template_name = 'page_maker/template_delete.html'
```
PATHLIB

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    slug_field = 'name'
+    slug_url_kwarg = 'templatename'
+
+    def test_func(self):
```
See my comments on previous `test_func`s

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+
+
+class CommentCreateView(LoginRequiredMixin, FormView):
+    http_method_names = ['post']  # POST request only
```
Maybe capitalize `POST` in the method names? (Don't know how Django handles this...) But the comment is probably extraneous.

tameTNT replied:
Not sure; @baru64 ?

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    http_method_names = ['post']
+    model = Comment
+
+    def test_func(self):
```
See my comments on previous `test_func`s

A5rocks commented on jolly-jellyfish/src/static/style.css:
```diff
@@ -0,0 +1,102 @@
+    border-right: 2px solid black;
+    border-bottom: 2px solid black;
+    padding: 3px;
+    font-family: hellovetica, "MS Sans Serif", "News Cycle", sans-serif;
```
Shouldn't this inherit from `body`?

Artemis21 replied:
No, buttons are not inheriting fonts either on my browser.

A5rocks commented on jolly-jellyfish/src/static/style.css:
```diff
@@ -0,0 +1,102 @@
+    border-right: 2px solid white;
+    border-bottom: 2px solid white;
+    padding: 3px;
+    font-family: hellovetica, "MS Sans Serif", "News Cycle", sans-serif;
```
Shouldn't this inherit from `body`?

Artemis21 replied:
No, the contents of input fields don't inherit fonts.

A5rocks reviewed (changes requested):
Most of these are purely appearance-based... But :/

God was this a long PR...

tameTNT committed:
Update jolly-jellyfish/README.md pipenv title

Clarity update in prep for potential future changes.

Co-authored-by: EXPLOSION <40616000+A5rocks@users.noreply.github.com>

tameTNT commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+
 ## Initial Setup
-```sh
+### wkhtmltoimage
````
```suggestion
### Selenium
```

tameTNT commented on jolly-jellyfish/README.md:
```diff
@@ -1,18 +1,52 @@
+Please also note that, as described within `src/page_maker/templatetags/render_thumbnail.py`, 
+asynchronous tasks (using `Celery` and `Redis`) are not easily setup on Windows - 
+hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
+
```
```suggestion
FIXME: find if we can use docker for celery / redis

```

tameTNT commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
+
+### pipenv
+```shell script
````
```suggestion
```sh
```

tameTNT commented on jolly-jellyfish/README.md:
````diff
@@ -1,18 +1,52 @@
+
+### pipenv
+```shell script
 $ pipenv install
````
```suggestion
$ cd jolly-jellyfish
$ pipenv install
```

tameTNT commented on jolly-jellyfish/src/django_meta/settings.py:
```diff
@@ -4,9 +4,14 @@
 # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
 BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 MEDIA_ROOT = os.path.join(BASE_DIR, 'user_media')
+MEDIA_URL = "/user_media/"
```
```suggestion
MEDIA_URL = '/user_media/'
```
Consistency improvement.

tameTNT commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,39 @@
+    # Using RelativeFilePathField allows for using django static support easily within the template
     # match uses regex applied to base filename only(.*=any characters, \.=., $=end of string)
-    style_sheet = models.FilePathField(path=get_theme_path, recursive=True, match='.*\.css$')
+    style_sheet = RelativeFilePathField(path=get_theme_path, recursive=True, match='.*\.css$')
```
```suggestion
    style_sheet = RelativeFilePathField(path=get_theme_path, recursive=True, match='.+\.css$')
```
Prevent uploads of a file named '.css'

tameTNT committed:
Update webpage_create.html with energy!

Bit more exciting this way!

Co-authored-by: EXPLOSION <40616000+A5rocks@users.noreply.github.com>

tameTNT committed:
Vastly improve readability of render_thumbnail.py

Thanks to @A5rocks for all the insight on the PR!
The file, esp. variable names, is now much more readable.
(and it also actually supports Firefox rather than just pretending to)

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

tameTNT commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+)
+
+urlpatterns = [
+    path('', MainView.as_view(), name='home'),
+    path('users/register', UserRegister.as_view(), name='register'),
```
```suggestion
    path('/', MainView.as_view(), name='home'),
    path('users/register/', UserRegister.as_view(), name='register'),
```
Yes, technically, the other two should end with / too.

tameTNT commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+    path('pages/<str:pagename>/', WebpageView.as_view(), name='webpage-view'),
+    path('pages/<str:pagename>/detail', WebpageDetailView.as_view(), name='webpage-detail'),
+    path('pages/<str:pagename>/update', WebpageUpdateView.as_view(), name='webpage-update'),
+    path('pages/<str:pagename>/delete', WebpageDeleteView.as_view(), name='webpage-delete'),
```
```suggestion
    path('pages/<str:pagename>/detail/', WebpageDetailView.as_view(), name='webpage-detail'),
    path('pages/<str:pagename>/update/', WebpageUpdateView.as_view(), name='webpage-update'),
    path('pages/<str:pagename>/delete/', WebpageDeleteView.as_view(), name='webpage-delete'),
```

tameTNT commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+    path('pages/<str:pagename>/delete', WebpageDeleteView.as_view(), name='webpage-delete'),
+    path('pages/', WebpageListView.as_view(), name='webpage-list'),
+
+    path('pages/<str:pagename>/comment', CommentCreateView.as_view(), name='comment-create'),
+    path('comments/<int:pk>/delete', CommentDeleteView.as_view(), name='comment-delete'),
+
+    path('templates/new', TemplateCreateView.as_view(), name='template-create'),
```
```suggestion
    path('pages/<str:pagename>/comment/', CommentCreateView.as_view(), name='comment-create'),
    path('comments/<int:pk>/delete/', CommentDeleteView.as_view(), name='comment-delete'),

    path('templates/new/', TemplateCreateView.as_view(), name='template-create'),
```

tameTNT commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -0,0 +1,42 @@
+
+    path('templates/new', TemplateCreateView.as_view(), name='template-create'),
+    path('templates/<str:templatename>/', TemplateView.as_view(), name='template-view'),
+    path('templates/<str:templatename>/delete', TemplateDeleteView.as_view(), name='template-delete'),
```
```suggestion
    path('templates/<str:templatename>/delete/', TemplateDeleteView.as_view(), name='template-delete'),
```

tameTNT reviewed (commented):
That's a lot of comments! Thanks again @A5rocks in like the 3rd communication channel (Discord, commit message and these comments!)

tameTNT commented:
Also, I'm unsure as to whether this is actually the case or if just passing executable_path to the webdriver.Chrome/Firefox instance is enough (below):
`driver = webdriver.Chrome(executable_path=DRIVER_PATH, options=options)`

https://pypi.org/project/selenium/
> Drivers
> Selenium requires a driver to interface with the chosen browser. Firefox, for example, requires geckodriver, which needs to be installed before the below examples can be run. Make sure it’s in your PATH, e. g., place it in /usr/bin or /usr/local/bin.
> 
> Failure to observe this step will give you an error selenium.common.exceptions.WebDriverException: Message: ‘geckodriver’ executable needs to be in PATH.

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/example_page.html:
```diff
@@ -0,0 +1,15 @@
+{% extends "page_maker/base.html" %}
+
+{% block content %}
+<h1><span>Example Page title</span></h1>
```
This is required to be `<h1><span><span>Example Page title</span></span></h1>` by tht hotdog theme.

tameTNT replied:
```suggestion
<h1><span><span>Example Page title</span></span></h1>
```

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/example_page.html:
```diff
@@ -0,0 +1,15 @@
+    Example Subtitle paragraph
+<p>
+    {% load static %}
+    <img src="{% static "images/placeholder_img.png" %}" alt="Example image">
```
This is required to be `<div class="img-wrapper"><img ...></div>` by several of the themes.

tameTNT replied:
```suggestion
    <div class="img-wrapper"><img src="{% static "images/placeholder_img.png" %}" alt="Example image"></div>
```

Artemis21 commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -1,3 +1,219 @@
+    template_name = 'page_maker/webpage_create.html'
+    form_class = WebpageForm
+
+    # TODO validate and sanitize text and images
```
How much of this does django already do for us?

A5rocks replied:
It's been updated:
`     # NB: Django pre-sanitizes text input already - https://docs.djangoproject.com/en/dev/topics/security/`

Artemis21 reviewed (commented):
Looks good mostly, some changes needed for styling. Don't indent after `{% csrf_token %}`, it's a single statement rather than the opening of a block.

tameTNT committed:
Apply suggestions from code review - README.md

README.md changes for clarity - suggested on PR code review

tameTNT committed:
Apply suggestions from code review

Clarity and consistency improvements in models.py, urls.py and settings.py along with a correction in README.md

tameTNT commented on jolly-jellyfish/src/page_maker/models.py:
```diff
@@ -1,26 +1,37 @@
+    style_sheet = RelativeFilePathField(path=get_theme_path, recursive=True, match='.+\.css$')
+
+    # null=False and blank=False enforce there always being a thumbnail image associated
+    thumbnail = models.ImageField(null=False, blank=False, editable=False
```
```suggestion
    thumbnail = models.ImageField(null=False, blank=False, editable=False,
```

tameTNT committed:
Re-add missing comma

Trust me to break things when adding suggestions.

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/main.html:
```diff
@@ -0,0 +1,8 @@
+<a href="{% url 'webpage-create' %}">create page</a>
+<a href="{% url 'template-create' %}">create template</a>
+<a href="{% url 'webpage-list' %}">best pages</a>
+{% endblock %}
```
```suggestion
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/template_create.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+  <form method="POST" enctype="multipart/form-data">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Create template</button>
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
    {{ form.as_p }}
    <button type="submit">Create template</button>
  </form>
</div>
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/template_delete.html:
```diff
@@ -0,0 +1,15 @@
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <p>Delete template <b>{{template.name}}</b></p>
+      <button type="submit">Confirm</button>
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
    {{ form.as_p }}
    <p>Delete template <b>{{template.name}}</b></p>
    <button type="submit">Confirm</button>
  </form>
</div>
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_delete.html:
```diff
@@ -0,0 +1,15 @@
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <p>Delete account <b>{{user.username}}</b></p>
+      <button type="submit">Confirm</button>
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
    {{ form.as_p }}
    <p>Delete account <b>{{user.username}}</b></p>
    <button type="submit">Confirm</button>
  </form>
</div>
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -0,0 +1,27 @@
+<a href="{% url 'user-delete' pk=viewed_user.id %}">delete</a>
+{% endif %}
+<div>
+  <table>
```
```suggestion
  <table>
  <tr>
    <th>Preview</th>
    <th>Page Name</th>
  </tr>
```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_update.html:
```diff
@@ -0,0 +1,14 @@
+
+{% block content %}
+<h2>Update profile</h2>
+<div>
+  <form method="POST">
+  {% csrf_token %}
+    <div>
+      {{ form.as_p }}
+      <button type="submit">Update</button>
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
<form method="POST">
{% csrf_token %}
    {{ form.as_p }}
    <button type="submit">Update</button>
</form>
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_delete.html:
```diff
@@ -0,0 +1,15 @@
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_update.html:
```diff
@@ -0,0 +1,14 @@
+    </div>
+  </form>
+</div>
+{% endblock %}
```
```suggestion
{% endblock %}

```

tameTNT commented on jolly-jellyfish/src/page_maker/templates/registration/login.html:
```diff
@@ -0,0 +1,25 @@
+         <input type="submit" value="login" />
+         <input type="hidden" name="next" value="{{ next }}" />
+     </form>
+{% endblock %}
```
```suggestion
{% endblock %}

```

tameTNT committed:
Apply suggestions from code review - misc. html formatting

Polish HTML formatting including newlines at end-of-file and some typo and sylting corrections.

tameTNT committed:
Neaten HTML templates

Adjust indentation across templates to 4 spaces (in line with Python).
Remove functionless <div> wrappers.
Add newlines (inc. to EOF) and more spacing to cluttered files.

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Improve formatting and clarity

Remove redundant arguments in super() function
Add detail to comments

tameTNT committed:
Apply suggestions from code review - add trailing '/'

Improve consistency across views.py file

tameTNT reviewed (commented):
Applied further changes from comments. Finish discussing any remaining ones and then this branch can be merged.

tameTNT committed:
Tidy TODOs and remove extraneous comments

Add new packages needed to requirements.txt
Add minor formatting changes to example_page.html

tameTNT committed:
Merge remote-tracking branch 'origin/like-feature-(and-page-ranking)' into like-feature-(and-page-ranking)

tameTNT committed:
Add more num likes labels and enable pagination

Edit urls.py to conform with no leading '/' convention.
Actually implement pagination code on webpage_list.html with buttons to
go to the next/previous page etc.

tameTNT committed:
Improve basic page functionality

Add '?next={{ request.path }}' to end of urls to ensure user is
redirected back to the page where they clicked the button. More of a
convenience feature than anything!

Added clarification to user_detail.html so that the page says if the
user hasn't created any pages.

Corrected typo in register.html.

tameTNT committed:
Fully add 'like' feature to relevant pages

Users can now, if logged in, like/unlike pages (but not their own!)
A webpage's like count is now displayed in various locations - including
on the 'best pages' page and the page itself (webpage_detail.html).

Minor sorting improvements in UserDetailView to sort by edit datetime.

tameTNT committed:
Add a 'recent pages' page

New url for recent pages, 'pages-recent/', and by extension for 'best'
also,'pages-top/'. These are both WebpageListView views, just with
different ordering values.
Update main.html to show this new navigation option.

A5rocks commented:
A list of unresolved discussions for easy reference without having to scroll through all the above:

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r465967672

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r465976901

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r465985028 (replicated in other places too)

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r466501932

Artemis21 committed:
Merge pull request #23 from Artemis21/use-platform-paths

Use Pathlib for paths for cross platform compability.

A5rocks commented on jolly-jellyfish/README.md:
````diff
@@ -34,17 +34,21 @@ Therefore, only one executable would need to be download to the src directory an
+### Using pipenv
+```sh
+$ cd src
+$ cd jolly-jellyfish
````
Whoops, ordered incorrectly: either remove `cd jolly-jellyfish` or reverse the ordering.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/base.html:
```diff
@@ -34,11 +34,11 @@
-    <a href="{% url 'register' %}">Register</a>
+        <a href="{% url 'login' %}">Login</a>
+        <a href="{% url 'register' %}">Register</a>
     {% endif %}
```
I think we're going with the indentation style that goes:

```
{% if user.is_authenticated %}
<a href="{% url 'user-detail' pk=user.id %}">{{ user.username }}</a>
<a href="{% url 'logout' %}">Logout</a>
{% else %}
<a href="{% url 'login' %}">Login</a>
<a href="{% url 'register' %}">Register</a>
{% endif %}
```
instead of what's already here.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html:
```diff
@@ -3,25 +3,28 @@
+                <img src="{{ webpage.thumbnail.url }}">
+            {% else %}
+                Go to page itself to generate a thumbnail for it.
+            {% endif %}
```
I think we decided to not indent inside of if blocks.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -4,26 +4,26 @@
     {% if webpage.user_image_1 %}
-    <img src="{{ webpage.user_image_1.url }}" alt="User-uploaded header image">
+        <img src="{{ webpage.user_image_1.url }}" alt="User-uploaded header image">
     {% endif %}
```
I don't think we're indenting inside of if blocks.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -4,26 +4,26 @@
     {% if webpage.user_image_3 %}
-    <img src="{{ webpage.user_image_3.url }}" alt="User-uploaded footer image">
+        <img src="{{ webpage.user_image_3.url }}" alt="User-uploaded footer image">
     {% endif %}
```
I don't think we're indenting inside of if blocks.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -4,26 +4,26 @@
     {% if webpage.user_image_2 %}
-    <img src="{{ webpage.user_image_2.url }}" alt="User-uploaded main image">
+        <img src="{{ webpage.user_image_2.url }}" alt="User-uploaded main image">
     {% endif %}
```
I don't think we're indenting inside if blocks.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -23,9 +24,18 @@
+    :return Options object with common arguments added
+    """
+    options_object.add_argument("--headless")
+    options_object.add_argument("--hide-scrollbars")
```
Double-quotes -> single-quotes.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -52,40 +62,42 @@ def render_thumbnail(request: HttpRequest, url_to_render: str, page_obj: Union[W
         # 'template_' prefix for Template Objects; 'pg_' for Webpages.
         # Can't use only the obj's name as this may contain characters not permitted by file system
-        image_path = Path('thumbnails') / f'{"template_" if template_obj_bool else "pg_"}thumb-{str(page_obj.id)}.png'
+        image_path = Path('thumbnails') / f'{"template_" if is_template else "pg_"}thumb-{str(page_obj.id)}.png'
```
These are forced double-quotes I believe, but maybe we can move this if statement up into an `image_prefix` variable or something.

Also, I believe f-strings automatically call `str` on the arguments unless you do `f'{5!r}'` (for example) in which it gets the `repr`.

A5rocks commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -94,23 +106,25 @@ def render_thumbnail(request: HttpRequest, url_to_render: str, page_obj: Union[W
 
         img = Image.open(full_os_path)
-        img.thumbnail(size=(500, 500))  # image shrunk to reduce storage space and reduce space used on actual website
+        # image shrunk (aspect ration maintained) to reduce storage space and reduce  footprint used on actual webpage
```
Double space before `footprint`.

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -20,23 +20,23 @@
+    path('pages/<str:pagename>/detail/', WebpageDetailView.as_view(), name='webpage-detail'),
+    path('pages/<str:pagename>/update/', WebpageUpdateView.as_view(), name='webpage-update'),
+    path('pages/<str:pagename>/delete/', WebpageDeleteView.as_view(), name='webpage-delete'),
     path('pages/', WebpageListView.as_view(), name='webpage-list'),
```
Pretty sure our convention here is to remove the `/` (at least, that's what baru is doing in his PR)

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -20,23 +20,23 @@
+    path('templates/new/', TemplateCreateView.as_view(), name='template-create'),
     path('templates/<str:templatename>/', TemplateView.as_view(), name='template-view'),
-    path('templates/<int:pk>/delete', TemplateDeleteView.as_view(), name='template-delete'),
+    path('templates/<str:templatename>/delete/', TemplateDeleteView.as_view(), name='template-delete'),
```
Same as above - I just don't have a big enough screen to select them all.

A5rocks reviewed (changes requested):
A whole lot of stylistic things / convention things.

~~By the way, your HTML editor is improperly set up and is putting 4-space indentation instead of 2-space :P~~

I should have commented on all the spots though.

A5rocks commented:
Oh wait oops, looking into it it seems 4-space style was already in the files... Well, I'll delete all of those comments (if I can, resolve if not), and then look into an html formatter to make sure everything is consistent :)

A5rocks committed:
Fix `Using pipenv` section

A5rocks committed:
Make code consistent

There were several things where this wasn't consistent, such as:
 - 4-space indentation for HTML, 2-space is convention (I think)
 - Indentation within if blocks (this is broken intentionally in the one place that remains)
 - Single-quotes instead of double-quotes
 - URLs shouldn't end with `/`

A5rocks commented:
Everything should be resolved...

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r465985028 is out of scope for this PR

https://github.com/Artemis21/summer-code-jam-2020/pull/24#discussion_r465976901 is also out of scope for this PR

A5rocks committed:
Merge branch 'add-page-render-functionality' into like-feature-(and-page-ranking)

A5rocks committed:
Make code more consistent

- Change HTML from 4-space indentation to 2-space indentation
- No indents in if blocks
- Prefer single-quotes unless the string has a contraction.

baru64 committed:
Remove divs from templates

tameTNT committed:
Change to Like and comment on my page

This additional text seems to have bee removed during one of the commit processes.

A5rocks committed:
More formatting of the HTML

A5rocks committed:
Fix for standard templates

The standard templates that ship with the app require doubly nested
spans (for spacing reasons...), so we need to make sure to supply these
for the Django templates.

A5rocks committed:
Merge pull request #28 from Artemis21/add-page_maker-app

Merge @baru64's feature branch

Artemis21 committed:
Remove config.json since it will be generated if missing.

Artemis21 committed:
Use Jan 1st, 0001 for dates in initial-themes.yaml.

Artemis21 committed:
Merge pull request #32 from Artemis21/add-docker-files

Add docker files

A5rocks committed:
Merge branch 'master' into add-page-render-functionality

A5rocks committed:
Merge branch 'add-page-render-functionality' into like-feature-(and-page-ranking)

A5rocks committed:
Merge pull request #29 from Artemis21/like-feature-(and-page-ranking)

Like feature and page ordering features

Artemis21 commented on jolly-jellyfish/src/config.json:
```diff
@@ -1,5 +0,0 @@
-{
-    "secret_key": null,
-    "debug": true,
-    "allowed_hosts": []
```
We *don't* want to delete this, right?

A5rocks replied:
Yeah.

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -5,18 +5,31 @@
-<img src="{{ webpage.user_image_3.url }}">
-<div>{{ webpage.user_text_3 }}</div>
-<img src="{{ webpage.user_image_4.url }}">
+<h1><span>{{ webpage.user_title }}</span></h1>
```
```suggestion
<h1><span><span>{{ webpage.user_title }}</span></span></h1>
```
Missing a level of spans

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -5,18 +5,31 @@
 <p>
-  <a href="{% url 'webpage-detail' pagename=webpage.name %}">Comment on my page</a>
+  {% if webpage.user_image_1 %}
+  <img src="{{ webpage.user_image_1.url }}" alt="User-uploaded header image">
```
```suggestion
  <div class="img-wrapper"><img src="{{ webpage.user_image_1.url }}" alt="User-uploaded header image"></div>
```

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -5,18 +5,31 @@
+<p>{{ webpage.user_text_1 }}</p>
+<p>
+  {% if webpage.user_image_2 %}
+  <img src="{{ webpage.user_image_2.url }}" alt="User-uploaded main image">
```
```suggestion
  <div class="img-wrapper"><img src="{{ webpage.user_image_2.url }}" alt="User-uploaded main image"></div>
```

Artemis21 commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -5,18 +5,31 @@
+<p>{{ webpage.user_text_2 }}</p>
+<p>
+  {% if webpage.user_image_3 %}
+  <img src="{{ webpage.user_image_3.url }}" alt="User-uploaded footer image">
```
```suggestion
  <div class="img-wrapper"><img src="{{ webpage.user_image_3.url }}" alt="User-uploaded footer image"></div>
```

Artemis21 commented on jolly-jellyfish/src/page_maker/templatetags/render_thumbnail.py:
```diff
@@ -0,0 +1,131 @@
+            driver = webdriver.Firefox(executable_path=DRIVER_PATH, options=options)
+        else:
+            raise Exception('The driver specified in SELENIUM_DRIVER is not supported.\n'
+                            "Currently, only Chrome ('chromedriver') and Firefox ('geckodriver') are supported.\n"
```
```suggestion
                            "Currently, only Chrome/Chromium ('chromedriver') and Firefox ('geckodriver') are supported.\n"
```
Chromedriver also works for Chromium.

A5rocks committed:
Apply suggestions from @Artemis21

Mostly to do with integrating with the standard templates.

Co-authored-by: Artemis <hollyshort21@gmail.com>

A5rocks committed:
Re-add default configuration file

Before, you had to have some knowledge of the source to change
allowed_hosts... Not anymore :)

Artemis21 reviewed (approved):
LGTM
